### PR TITLE
[bitnami/mongodb] support https://github.com/combor/k8s-mongo-labeler-sidecar - a very …

### PR DESCRIPTION
**Pull request by korenlev**
…crucial component for allowing WRITE access to the mongodb replica-set

**support https://github.com/combor/k8s-mongo-labeler-sidecar - a very crucial component for allowing WRITE access to the MongoDB replica-set"**

**External access to the replica-set dynamic primary pod**

**No drawbacks, tested**

**Applicable to common need to allow external access to MongoDB primary pod**

  - fixes # https://github.com/bitnami/charts/issues/6434

**Checklist** 

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)

### CODE REVIEW

**Comment by korenlev** [comment]
instead of adding a whole new section to mongodb main values.yaml i have suggested a new values.primary-mongodb-labeler-sidecar.yaml that includes a clear diff (additional on top of the current values.yaml)

**Comment by alemorcuq** [comment]
Hi, @korenlev. I think this is a very specific feature and I feel like it will be better to support an `extraDeploy` to allow users deploy any extra objects they want, like we do in other charts (for example: https://github.com/bitnami/charts/blob/master/bitnami/redis-cluster/values.yaml#L758). What do you think?

**Comment by korenlev** [comment]
Hi alejandro, isn't it common to have a way to connect to the primary as
the only one allowing WRITE to mongodb ?

בתאריך יום ו׳, 21 במאי 2021, 10:11, מאת Alejandro Moreno ‏<
***@***.***>:

**Comment by korenlev** [comment]
Hi alejandro, isn't it common to have a way to connect to the primary as
the only one allowing WRITE to mongodb ?
Currently, on this bitnami chart, any client connecting to it needs to
figure out (periodically, to make sure) who is primary, then if he is
connecting from outaide the cluster, setup his own service (agaib,
periodically when primary pod have moved).
Seems like a great deal to me, making current deployment very hard to use

בתאריך שבת, 22 במאי 2021, 19:45, מאת Koren Lev ***@***.***>:

**Comment by korenlev** [comment]
i believe  you notes: "without using a service that is load balancing
between the pods" - is the whole point, and actually is the missing piece
from the chart.
as there is only 1 PRIMARY POD and only this PRIMARY POD can allow WRITE
access, it needs to have a 'special service' that is 'outside' any
load-balancing , load-balancing is only good for READ from replica-set.
am i wrong ? i can drop it as extraDeploy no worries ...

‫בתאריך יום ג׳, 25 במאי 2021 ב-17:54 מאת ‪Alejandro Moreno‬‏ <‪
***@***.***‬‏>:‬

**Comment by korenlev** [comment]
moving to contribute the extra-deploy[] method suggested ...

**Comment by korenlev** [comment]
done adding as extraDeploy feature - please review

**Comment by alemorcuq** [comment]
> i believe  you notes: "without using a service that is load balancing
> between the pods" - is the whole point, and actually is the missing piece
> from the chart.

As I said earlier, the `externalAccess` feature in the chart creates one service per pod, so you can connect to MongoDB using a string with all the services instead of just one service providing load balancing to the pods. Please, check that feature as I think it solves your problem.

**Comment by korenlev** [comment]
Hi Alejandro,

externalAccess exposes ALL replica pods through services.
This does NOT solve the problem of connecting to the PRIMARY POD of the
replica set, if someone wishes to WRITE only.
"using a string" - maybe you mean using replica URL l- this does NOT solve
the problem either as the load-balancing hits SECONDARY pods once in a
while where only READ is allowed.
This is a big problem to solve in the current chart, i have no idea how one
can use it for WRITE access without this solution.

can you please test and take a closer look?

BR
Koren

‫בתאריך יום ב׳, 31 במאי 2021 ב-17:15 מאת ‪Alejandro Moreno‬‏ <‪
***@***.***‬‏>:‬

**Comment by juan131** [comment]
Hi Koren,

Thanks so much for the contribution!!

As you said, our current approach to expose MongoDB consist on creating a LoadBalancer or NodePort service per MongoDB node (regardless this node is a primary or secondary node). This approach has two downsides:

- Logic must be implemented in the client side to dynamically detect which is the primary node, so "write" actions aren't sent to the secondary nodes. 
- It's very expensive (when using LoadBalancer svc type) since you need to request to the CloudProvider several external IPs which will increase your operational cost.

Ideally, we should have some kind of intelligent proxy between the client and MongoDB nodes with the ability to identify "write" queries and redirect them to the primary node, while load-balancing "read" queries using round-robin. This will be the optimal solution IMO.

The approach the "MongoDB replica set pod labeler" uses is a possible workaround, although not a very "clean" solution IMO. Assigning labels dynamically to the pods and giving RBAC permissions to a sidecar to do so sounds risky to me. That's why I have my concerns regarding officially support it in our chart.

The solution @alemorcuq proposed you about using the **"extraDeploy"** to use the"MongoDB replica set pod labeler" doesn't require so many changes. He simply assumed we already had support this **"extraDeploy"** parameter (which is very common in our charts) but it we hadn't added it to MongoDB yet. Let me illustrate it using an example using the current chart:

- Ensure you have the changes I just submitted at https://github.com/bitnami/charts/pull/6536 (they add support for **"extraDeploy"** parameter)
- Create  a **values.yaml** like the one below:

```yaml
architecture: replicaset
replicaCount: 3
sidecars:
  - name: mongo-labeler
    image: korenlev/k8s-mongo-labeler-sidecar
    imagePullPolicy: Always
    env:
      - name: LABEL_SELECTOR
        value: "app.kubernetes.io/component=mongodb,app.kubernetes.io/instance=mongodb,app.kubernetes.io/name=mongodb"
      - name: NAMESPACE
        value: "default"
      - name: DEBUG
        value: "true"
extraDeploy:
  - apiVersion: v1
    kind: Service
    metadata:
      name: mongodb-primary
      namespace: default
      labels:
        app.kubernetes.io/component: mongodb
        app.kubernetes.io/instance: mongodb
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: mongodb
    spec:
      type: LoadBalancer
      externalTrafficPolicy: Cluster
      ports:
        - name: mongodb
          port: 27017
          protocol: TCP
          targetPort: mongodb
      selector:
        app.kubernetes.io/component: mongodb
        app.kubernetes.io/instance: mongodb
        app.kubernetes.io/name: mongodb
        primary: "true"
```

- Then, install the MongoDB chart in the "default" namespace:

```console
$ helm install mongodb bitnami/mongodb -f values.yaml
```

**Comment by korenlev** [comment]
thanks Juan , got it all and i'll be working on it

**Comment by korenlev** [comment]
@juan131 i have merged your changes and added the extraDeploy and needed section , can you verify please

**Comment by korenlev** [comment]
@juan131 i have tested this with your changes (only uncommenting sections added)  in production - results looks fine, primary pod exposed as designed

File: bitnami/mongodb/README.md [hunk 1]
```diff
@@ -531,3 +531,34 @@ Use the workaround below to upgrade from versions previous to 5.0.0. The followi
 ```console
 $ kubectl delete statefulset my-release-mongodb-arbiter my-release-mongodb-primary my-release-mongodb-secondary --cascade=false
 ```
+
+### Add extra deployment options
+
+To add extra deployments (useful for advanced features like sidecars), use the `extraDeploy` property.
+# example here for a 'mongo-labeler' sidecar to find primary pod in replica set and listen on predictable port
```
Comment (juan131) [inline bitnami/mongodb/README.md:538]: Please note that, by adding this `# ...` paragraph, you're include a H1 header:
```suggestion

In the example below, you can find how to use a example here for a [MongoDB replica set pod labeler sidecar](https://github.com/combor/k8s-mongo-labeler-sidecar) to identify the primary pod and dynamically label it as the primary node:
```

File: bitnami/mongodb/README.md [hunk 2]
```diff
@@ -531,3 +531,34 @@ Use the workaround below to upgrade from versions previous to 5.0.0. The followi
...
+        - name: mongodb-primary
+          port: 30001
+          nodePort: 30001
+          protocol: TCP
+          targetPort: mongodb
+      selector:
+        app.kubernetes.io/component: mongodb
+        app.kubernetes.io/instance: mongodb
+        app.kubernetes.io/name: mongodb
+        primary: "true"
```
Comment (juan131) [inline bitnami/mongodb/README.md:564]: Close the "yaml" section block with three "`"

File: bitnami/mongodb/values.yaml [hunk 3]
```diff
@@ -443,6 +472,24 @@ initContainers: {}
...
+#- name: mongo-labeler
+#  image: korenlev/k8s-mongo-labeler-sidecar
+#  imagePullPolicy: Always
+#  env:
+#    - name: LABEL_SELECTOR
+#      value: "app.kubernetes.io/component=mongodb,app.kubernetes.io/instance=mongodb,app.kubernetes.io/name=mongodb"
+#    - name: NAMESPACE
+#      value: "the-mongodb-namespace"
+#    - name: DEBUG
+#      value: "true"
```
Comment (juan131) [inline bitnami/mongodb/values.yaml:490]: nice!

File: bitnami/mongodb/values.yaml [hunk 4]
```diff
@@ -52,6 +52,35 @@ clusterDomain: cluster.local
...
+#      - name: mongodb
+#        port: 30001
+#        nodePort: 30001
+#        protocol: TCP
+#        targetPort: mongodb
+#    selector:
+#      app.kubernetes.io/component: mongodb
+#      app.kubernetes.io/instance: mongodb
+#      app.kubernetes.io/name: mongodb
+#      primary: "true"
```
Comment (juan131) [inline bitnami/mongodb/values.yaml:81]: Nice, thanks!

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks @korenlev !! You changes LGTM, just a couple of suggestions for the **README.md**

File: bitnami/mongodb/README.md (hunk 2 above)
Comment (juan131) [reply bitnami/mongodb/README.md:565]: Still pending, right?

**Comment by juan131** [comment]
Please fix the "too many blank lines (1 > 0)  (empty-lines)" errors in the values.yaml that the linter is throwing

File: bitnami/mongodb/README.md (hunk 2 above)
Comment (korenlev) [reply bitnami/mongodb/README.md:565]: done now , sorry for the delay

**Review by juan131 — APPROVED** [review]
LGTM

**Comment by juan131** [comment]
Your changes were landed @korenlev

**Pull request merged**